File: packages/core/src/ini.ts
/**
 * Lossless INI reader/writer.
 *
 * Config files belong to the user and carry hand-written comments, so anything not understood passes through
 * untouched and setting one key rewrites exactly one line. Bytes are carried as latin1 strings: that mapping is
 * one-to-one and reversible in JavaScript, so a file in an unknown legacy codepage round-trips byte for byte
 * without ever being decoded as something it is not.
 */

import { latin1, latin1Bytes, splitLines } from "./text.js";

export type IniNode =
  | { kind: "blank"; raw: string }
  | { kind: "comment"; raw: string }
  | { kind: "section"; name: string; raw: string }
  | { kind: "entry"; section: string; key: string; value: string; raw: string; parts: EntryParts }
  | { kind: "unknown"; raw: string };

/** Pieces of an entry line, kept so a value change rewrites nothing but the value. */
interface EntryParts {
  indent: string;
  key: string;
  separator: string;
  /** Inline comment after the value, with its leading whitespace, e.g. " ; ORIGINAL 480". */
  comment: string;
  trailing: string;
  eol: string;
}

const SECTION_RE = /^(\s*)\[([^\]]*)\](\s*)$/;
const COMMENT_RE = /^\s*[;#]/;

// Whitespace as the regular-expression engine defines it, asked one character at a time. This file works in
// latin1, where 0xa0 is a non-breaking space, so a hand-written set would have to keep pace with that
// definition for the round-trip to stay byte-exact.
const SPACE_RE = /\s/;
const isSpace = (ch: string | undefined) => ch !== undefined && SPACE_RE.test(ch);

/**
 * Splits an entry line into its value and the pieces that surround it, or `undefined` when it is not an entry.
 *
 * Hand-parsed rather than matched by one expression: the natural expression needs a lazy key group in front of
 * the separator, and a backtracking engine walks a line that holds no "=" and ends in whitespace in quadratic
 * time - four times the length cost sixteen times the work, 60ms at 16k characters. Config files come from the
 * user and from mod archives, so their line lengths are not ours to bound.
 */
function parseEntry(body: string): { value: string; parts: Omit<EntryParts, "eol"> } | undefined {
  const eq = body.indexOf("=");
  if (eq < 0) return undefined;

  let keyStart = 0;
  while (keyStart < eq && isSpace(body[keyStart])) keyStart++;
  let keyEnd = eq;
  while (keyEnd > keyStart && isSpace(body[keyEnd - 1])) keyEnd--;
  // Nothing but whitespace in front of the "=" names no key, so the line is not an entry.
  if (keyEnd === keyStart) return undefined;

  // The separator carries the whitespace on both sides of the "=", so writing a value disturbs neither.
  let valueStart = eq + 1;
  while (isSpace(body[valueStart])) valueStart++;

  let trailingStart = body.length;
  while (trailingStart > valueStart && isSpace(body[trailingStart - 1])) trailingStart--;

  // The value stops at an inline comment: the first run of whitespace that a ";" or "#" follows. Requiring the
  // whitespace keeps a separator inside a value (a path, a list) from being mistaken for one.
  let commentStart = trailingStart;
  for (let i = valueStart; i < trailingStart;) {
    if (!isSpace(body[i])) {
      i++;
      continue;
    }
    let after = i;
    while (isSpace(body[after])) after++;
    if (body[after] === ";" || body[after] === "#") {
      commentStart = i;
      break;
    }
    i = after;
  }

  return {
    value: body.slice(valueStart, commentStart),
    parts: {
      indent: body.slice(0, keyStart),
      key: body.slice(keyStart, keyEnd),
      separator: body.slice(keyEnd, valueStart),
      comment: body.slice(commentStart, trailingStart),
      trailing: body.slice(trailingStart),
    },
  };
}

const fold = (s: string) => s.toLowerCase();

export class IniDocument {
  private constructor(
    private readonly nodes: IniNode[],
    /** Terminator used for lines this document adds. */
    readonly dominantEol: string,
  ) {}

  static parse(text: string): IniDocument {
    const nodes: IniNode[] = [];
    let crlf = 0;
    let lf = 0;
    let section = "";

    for (const { body, eol } of splitLines(text)) {
      if (eol === "\r\n") crlf++;
      else if (eol === "\n") lf++;
      const raw = body + eol;

      if (body.trim() === "") {
        nodes.push({ kind: "blank", raw });
        continue;
      }
      if (COMMENT_RE.test(body)) {
        nodes.push({ kind: "comment", raw });
        continue;
      }
      const sec = SECTION_RE.exec(body);
      if (sec) {
        section = sec[2] ?? "";
        nodes.push({ kind: "section", name: section, raw });
        continue;
      }
      const parsed = parseEntry(body);
      if (parsed) {
        const { value, parts } = parsed;
        nodes.push({ kind: "entry", section, key: parts.key, value, raw, parts: { ...parts, eol } });
        continue;
      }
      nodes.push({ kind: "unknown", raw });
    }

    return new IniDocument(nodes, crlf >= lf && crlf > 0 ? "\r\n" : "\n");
  }

  /** Decode bytes as latin1 so the round-trip is byte-exact whatever the file's real encoding is. */
  static parseBytes(bytes: Uint8Array): IniDocument {
    return IniDocument.parse(latin1(bytes));
  }

  get(section: string, key: string): string | undefined {
    return this.findEntry(section, key)?.value;
  }

  /**
   * Every entry the file actually contains, with the comment block directly above it. sfall documents most of
   * its settings inline, so that comment is the only description available for keys the catalog does not model.
   */
  entries(): Array<{ section: string; key: string; value: string; comment?: string }> {
    const out: Array<{ section: string; key: string; value: string; comment?: string }> = [];
    for (const [i, n] of this.nodes.entries()) {
      if (n.kind !== "entry") continue;
      const comment: string[] = [];
      for (let j = i - 1; j >= 0; j--) {
        const prev = this.nodes[j];
        if (prev?.kind !== "comment") break;
        const text = prev.raw.replace(/^\s*[;#]\s?/, "").trimEnd();
        // A run of X characters is sfall's section divider, not documentation.
        if (/^X{6,}$/.test(text.trim())) break;
        comment.unshift(text);
      }
      out.push({
        section: n.section,
        key: n.key,
        value: n.value,
        ...(comment.length ? { comment: comment.join(" ").trim() } : {}),
      });
    }
    return out;
  }

  /** Every section present in the file, in order of first appearance. */
  sections(): string[] {
    const seen: string[] = [];
    for (const n of this.nodes) if (n.kind === "section" && !seen.includes(n.name)) seen.push(n.name);
    return seen;
  }

  /**
   * Write a value, rewriting one line and only when it actually differs. A missing key is appended to its
   * section; a missing section is appended to the file.
   */
  set(section: string, key: string, value: string): void {
    const existing = this.findEntry(section, key);
    if (existing) {
      if (existing.value === value) return;
      const p = existing.parts;
      existing.value = value;
      // The inline comment is carried across: it is the author's note, not part of the value.
      existing.raw = p.indent + p.key + p.separator + value + p.comment + p.trailing + p.eol;
      return;
    }

    const eol = this.dominantEol;
    const entry: IniNode = {
      kind: "entry",
      section,
      key,
      value,
      raw: `${key}=${value}${eol}`,
      parts: { indent: "", key, separator: "=", comment: "", trailing: "", eol },
    };

    const at = this.endOfSection(section);
    if (at === -1) {
      this.terminateLine(this.nodes.length - 1);
      this.nodes.push({ kind: "section", name: section, raw: `[${section}]${eol}` }, entry);
    } else {
      // The line we insert after need not be the file's last, but it can be - config files are not required to
      // end with a newline, and splicing after an unterminated line would join two keys into one.
      this.terminateLine(at - 1);
      this.nodes.splice(at, 0, entry);
    }
  }

  toString(): string {
    return this.nodes.map((n) => n.raw).join("");
  }

  toBytes(): Uint8Array {
    return latin1Bytes(this.toString());
  }

  private findEntry(section: string, key: string) {
    for (const n of this.nodes) {
      if (n.kind === "entry" && fold(n.section) === fold(section) && fold(n.key) === fold(key)) return n;
    }
    return undefined;
  }

  /** Index just past the last meaningful line of a section, or -1 when the section is absent. */
  private endOfSection(section: string): number {
    let inSection = false;
    let last = -1;
    for (const [i, n] of this.nodes.entries()) {
      if (n.kind === "section") {
        if (inSection) break;
        inSection = fold(n.name) === fold(section);
        if (inSection) last = i + 1;
        continue;
      }
      if (inSection && n.kind !== "blank") last = i + 1;
    }
    return inSection ? last : -1;
  }

  /** Give the node at `index` a line terminator if it lacks one, so the next line starts on its own. */
  private terminateLine(index: number): void {
    const node = this.nodes[index];
    if (node && node.raw !== "" && !node.raw.endsWith("\n")) node.raw += this.dominantEol;
  }
}

File: crates/core/src/ini.rs
//! Lossless INI reader and writer.
//!
//! Config files belong to the user and carry hand-written comments, so anything not understood
//! passes through untouched and setting one key rewrites exactly one line. Content is carried as
//! bytes, so a file in an unknown legacy codepage round-trips byte for byte without ever being
//! decoded as something it is not.

use crate::text::{Line, fold_byte, is_space, latin1, same_name, split_lines};

/// Pieces of an entry line, kept so a value change rewrites nothing but the value.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
struct EntryParts {
    indent: Vec<u8>,
    key: Vec<u8>,
    separator: Vec<u8>,
    /// Inline comment after the value, with its leading whitespace, e.g. `" ; ORIGINAL 480"`.
    comment: Vec<u8>,
    trailing: Vec<u8>,
    eol: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum Node {
    Blank {
        raw: Vec<u8>,
    },
    Comment {
        raw: Vec<u8>,
    },
    Section {
        name: Vec<u8>,
        raw: Vec<u8>,
    },
    Entry {
        section: Vec<u8>,
        value: Vec<u8>,
        raw: Vec<u8>,
        parts: EntryParts,
    },
    Unknown {
        raw: Vec<u8>,
    },
}

impl Node {
    fn raw(&self) -> &[u8] {
        match self {
            Self::Blank { raw }
            | Self::Comment { raw }
            | Self::Section { raw, .. }
            | Self::Entry { raw, .. }
            | Self::Unknown { raw } => raw,
        }
    }

    fn raw_mut(&mut self) -> &mut Vec<u8> {
        match self {
            Self::Blank { raw }
            | Self::Comment { raw }
            | Self::Section { raw, .. }
            | Self::Entry { raw, .. }
            | Self::Unknown { raw } => raw,
        }
    }
}

/// One entry as a reader sees it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Entry {
    pub section: String,
    pub key: String,
    pub value: String,
    /// The comment block directly above the entry. sfall documents most of its settings inline, so
    /// this is the only description available for keys the catalog does not model.
    pub comment: Option<String>,
}

fn is_comment(body: &[u8]) -> bool {
    body.iter()
        .find(|&&b| !is_space(b))
        .is_some_and(|&b| b == b';' || b == b'#')
}

/// Splits a `[name]` line into its name, or `None` when the line is not a section header. The whole
/// line outside the brackets must be whitespace, and the name itself may not contain `]`.
fn parse_section(body: &[u8]) -> Option<&[u8]> {
    let open = body.iter().position(|&b| b == b'[')?;
    if !body[..open].iter().all(|&b| is_space(b)) {
        return None;
    }
    let rest = &body[open + 1..];
    let close = rest.iter().position(|&b| b == b']')?;
    if !rest[close + 1..].iter().all(|&b| is_space(b)) {
        return None;
    }
    Some(&rest[..close])
}

struct ParsedEntry {
    value: Vec<u8>,
    parts: EntryParts,
}

/// Splits an entry line into its value and the pieces that surround it, or `None` when it is not an
/// entry.
///
/// Hand-parsed rather than matched by one expression, as the TypeScript was: the natural expression
/// needs a lazy key group in front of the separator, and a backtracking engine walks a line holding
/// no `=` that ends in whitespace in quadratic time. Config files come from the user and from mod
/// archives, so their line lengths are not ours to bound.
fn parse_entry(body: &[u8]) -> Option<ParsedEntry> {
    let eq = body.iter().position(|&b| b == b'=')?;

    let mut key_start = 0;
    while key_start < eq && is_space(body[key_start]) {
        key_start += 1;
    }
    let mut key_end = eq;
    while key_end > key_start && is_space(body[key_end - 1]) {
        key_end -= 1;
    }
    // Nothing but whitespace in front of the "=" names no key, so the line is not an entry.
    if key_end == key_start {
        return None;
    }

    // The separator carries the whitespace on both sides of the "=", so writing a value disturbs
    // neither.
    let mut value_start = eq + 1;
    while value_start < body.len() && is_space(body[value_start]) {
        value_start += 1;
    }

    let mut trailing_start = body.len();
    while trailing_start > value_start && is_space(body[trailing_start - 1]) {
        trailing_start -= 1;
    }

    // The value stops at an inline comment: the first run of whitespace that a ";" or "#" follows.
    // Requiring the whitespace keeps a separator inside a value (a path, a list) from being mistaken
    // for one.
    let mut comment_start = trailing_start;
    let mut i = value_start;
    while i < trailing_start {
        if !is_space(body[i]) {
            i += 1;
            continue;
        }
        let mut after = i;
        while after < body.len() && is_space(body[after]) {
            after += 1;
        }
        if matches!(body.get(after), Some(b';' | b'#')) {
            comment_start = i;
            break;
        }
        i = after;
    }

    Some(ParsedEntry {
        value: body[value_start..comment_start].to_vec(),
        parts: EntryParts {
            indent: body[..key_start].to_vec(),
            key: body[key_start..key_end].to_vec(),
            separator: body[key_end..value_start].to_vec(),
            comment: body[comment_start..trailing_start].to_vec(),
            trailing: body[trailing_start..].to_vec(),
            eol: Vec::new(),
        },
    })
}

/// A run of six or more `X` characters is sfall's section divider, not documentation.
fn is_divider(text: &str) -> bool {
    let trimmed = text.trim();
    trimmed.len() >= 6 && trimmed.bytes().all(|b| b == b'X')
}

/// Strips one leading comment marker and at most one space after it, matching the TypeScript's
/// `/^\s*[;#]\s?/`.
fn comment_text(raw: &[u8]) -> String {
    let mut at = 0;
    while at < raw.len() && is_space(raw[at]) {
        at += 1;
    }
    if matches!(raw.get(at), Some(b';' | b'#')) {
        at += 1;
        if raw.get(at).copied().is_some_and(is_space) {
            at += 1;
        }
    }
    let mut end = raw.len();
    while end > at && is_space(raw[end - 1]) {
        end -= 1;
    }
    latin1(&raw[at..end])
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IniDocument {
    nodes: Vec<Node>,
    /// Terminator used for lines this document adds.
    dominant_eol: Vec<u8>,
}

impl IniDocument {
    /// Parses bytes. Never fails: a line that is not understood becomes an unknown node and passes
    /// through unchanged, which is the whole point of the reader.
    #[must_use]
    pub fn parse(text: &[u8]) -> Self {
        let mut nodes = Vec::new();
        let mut crlf = 0usize;
        let mut lf = 0usize;
        let mut section: Vec<u8> = Vec::new();

        for Line { body, eol } in split_lines(text) {
            if eol == b"\r\n" {
                crlf += 1;
            } else if eol == b"\n" {
                lf += 1;
            }
            let mut raw = body.to_vec();
            raw.extend_from_slice(eol);

            if body.iter().all(|&b| is_space(b)) {
                nodes.push(Node::Blank { raw });
                continue;
            }
            if is_comment(body) {
                nodes.push(Node::Comment { raw });
                continue;
            }
            if let Some(name) = parse_section(body) {
                section = name.to_vec();
                nodes.push(Node::Section {
                    name: name.to_vec(),
                    raw,
                });
                continue;
            }
            if let Some(parsed) = parse_entry(body) {
                let mut parts = parsed.parts;
                parts.eol = eol.to_vec();
                nodes.push(Node::Entry {
                    section: section.clone(),
                    value: parsed.value,
                    raw,
                    parts,
                });
                continue;
            }
            nodes.push(Node::Unknown { raw });
        }

        let dominant_eol = if crlf >= lf && crlf > 0 {
            b"\r\n".to_vec()
        } else {
            b"\n".to_vec()
        };
        Self {
            nodes,
            dominant_eol,
        }
    }

    /// The terminator lines added by [`IniDocument::set`] carry.
    #[must_use]
    pub fn dominant_eol(&self) -> &[u8] {
        &self.dominant_eol
    }

    /// The raw value of one key, or `None` when it is not present.
    #[must_use]
    pub fn get(&self, section: &str, key: &str) -> Option<&[u8]> {
        let at = self.find_entry(section.as_bytes(), key.as_bytes())?;
        match self.nodes.get(at) {
            Some(Node::Entry { value, .. }) => Some(value.as_slice()),
            _ => None,
        }
    }

    /// The value of one key decoded for display.
    #[must_use]
    pub fn get_str(&self, section: &str, key: &str) -> Option<String> {
        self.get(section, key).map(latin1)
    }

    /// Every entry the file actually contains, with the comment block directly above it.
    #[must_use]
    pub fn entries(&self) -> Vec<Entry> {
        let mut out = Vec::new();
        for (i, node) in self.nodes.iter().enumerate() {
            let Node::Entry {
                section,
                value,
                parts,
                ..
            } = node
            else {
                continue;
            };
            let mut comment: Vec<String> = Vec::new();
            for previous in self.nodes[..i].iter().rev() {
                let Node::Comment { raw } = previous else {
                    break;
                };
                let text = comment_text(raw);
                if is_divider(&text) {
                    break;
                }
                comment.push(text);
            }
            comment.reverse();
            let joined = comment.join(" ").trim().to_owned();
            out.push(Entry {
                section: latin1(section),
                key: latin1(&parts.key),
                value: latin1(value),
                comment: if joined.is_empty() {
                    None
                } else {
                    Some(joined)
                },
            });
        }
        out
    }

    /// Every section present in the file, in order of first appearance.
    #[must_use]
    pub fn sections(&self) -> Vec<String> {
        let mut seen: Vec<Vec<u8>> = Vec::new();
        for node in &self.nodes {
            if let Node::Section { name, .. } = node
                && !seen.iter().any(|held| same_name(held, name))
            {
                seen.push(name.clone());
            }
        }
        seen.iter().map(|name| latin1(name)).collect()
    }

    /// Writes a value, rewriting one line and only when it actually differs. A missing key is
    /// appended to its section; a missing section is appended to the file.
    pub fn set(&mut self, section: &str, key: &str, value: &[u8]) {
        if let Some(at) = self.find_entry(section.as_bytes(), key.as_bytes()) {
            let Some(Node::Entry {
                value: held,
                raw,
                parts,
                ..
            }) = self.nodes.get_mut(at)
            else {
                return;
            };
            if held == value {
                return;
            }
            held.clear();
            held.extend_from_slice(value);
            // The inline comment is carried across: it is the author's note, not part of the value.
            raw.clear();
            raw.extend_from_slice(&parts.indent);
            raw.extend_from_slice(&parts.key);
            raw.extend_from_slice(&parts.separator);
            raw.extend_from_slice(value);
            raw.extend_from_slice(&parts.comment);
            raw.extend_from_slice(&parts.trailing);
            raw.extend_from_slice(&parts.eol);
            return;
        }

        let eol = self.dominant_eol.clone();
        let mut raw = Vec::new();
        raw.extend_from_slice(key.as_bytes());
        raw.push(b'=');
        raw.extend_from_slice(value);
        raw.extend_from_slice(&eol);
        let entry = Node::Entry {
            section: section.as_bytes().to_vec(),
            value: value.to_vec(),
            raw,
            parts: EntryParts {
                key: key.as_bytes().to_vec(),
                separator: b"=".to_vec(),
                eol: eol.clone(),
                ..EntryParts::default()
            },
        };

        match self.end_of_section(section.as_bytes()) {
            None => {
                if !self.nodes.is_empty() {
                    self.terminate_line(self.nodes.len() - 1);
                }
                let mut header = Vec::new();
                header.push(b'[');
                header.extend_from_slice(section.as_bytes());
                header.push(b']');
                header.extend_from_slice(&eol);
                self.nodes.push(Node::Section {
                    name: section.as_bytes().to_vec(),
                    raw: header,
                });
                self.nodes.push(entry);
            }
            Some(at) => {
                // The line inserted after need not be the file's last, but it can be: config files
                // are not required to end with a newline, and splicing after an unterminated line
                // would join two keys into one.
                if at > 0 {
                    self.terminate_line(at - 1);
                }
                self.nodes.insert(at, entry);
            }
        }
    }

    /// Writes a value given as text, encoded back to latin1.
    pub fn set_str(&mut self, section: &str, key: &str, value: &str) {
        self.set(section, key, &crate::text::latin1_bytes(value));
    }

    #[must_use]
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut out = Vec::new();
        for node in &self.nodes {
            out.extend_from_slice(node.raw());
        }
        out
    }

    fn find_entry(&self, section: &[u8], key: &[u8]) -> Option<usize> {
        self.nodes.iter().position(|node| {
            matches!(node, Node::Entry { section: held, parts, .. }
                if same_name(held, section) && same_name(&parts.key, key))
        })
    }

    /// Index just past the last meaningful line of a section, or `None` when the section is absent.
    fn end_of_section(&self, section: &[u8]) -> Option<usize> {
        let mut in_section = false;
        let mut last = 0;
        for (i, node) in self.nodes.iter().enumerate() {
            if let Node::Section { name, .. } = node {
                if in_section {
                    break;
                }
                in_section = same_name(name, section);
                if in_section {
                    last = i + 1;
                }
                continue;
            }
            if in_section && !matches!(node, Node::Blank { .. }) {
                last = i + 1;
            }
        }
        in_section.then_some(last)
    }

    /// Gives the node at `index` a line terminator if it lacks one, so the next line starts on its
    /// own.
    fn terminate_line(&mut self, index: usize) {
        let eol = self.dominant_eol.clone();
        let Some(node) = self.nodes.get_mut(index) else {
            return;
        };
        let raw = node.raw_mut();
        if !raw.is_empty() && !raw.ends_with(b"\n") {
            raw.extend_from_slice(&eol);
        }
    }
}

/// Lowercases a name for use as a map key, where a borrowed comparison will not do.
#[must_use]
pub fn fold(name: &str) -> String {
    name.bytes().map(|b| fold_byte(b) as char).collect()
}
